File: app/src/main/java/com/dev/adflow/HomeScreen.kt
package com.dev.adflow

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.adflow.compose.AdFlowBanner
import com.adflow.compose.AdFlowNative
import com.adflow.core.AdFlow
import com.adflow.core.consent.PrivacyOptionsRequirement
import com.adflow.core.fullscreen.FullScreenCallback
import com.adflow.core.rewarded.RewardItem
import com.adflow.core.rewarded.RewardedAdCallback
import kotlinx.coroutines.launch

@Composable
fun HomeScreen() {
    val context = LocalContext.current
    var premium by remember { mutableStateOf(PremiumState.isPremium) }
    var lastReward by remember { mutableStateOf<RewardItem?>(null) }
    val snackbarHostState = remember { SnackbarHostState() }
    val scope = rememberCoroutineScope()
    val privacyOptionsRequired = AdFlow.consent.privacyOptionsRequirement == PrivacyOptionsRequirement.REQUIRED

    fun showBlocked(reason: String) {
        scope.launch { snackbarHostState.showSnackbar("Blocked: $reason") }
    }

    Scaffold(snackbarHost = { SnackbarHost(snackbarHostState) }) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .verticalScroll(rememberScrollState())
                .padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Row {
                Text("Premium (disable ads)")
                Switch(
                    checked = premium,
                    onCheckedChange = {
                        premium = it
                        PremiumState.isPremium = it
                    },
                )
            }

            Button(onClick = {
                (context as? android.app.Activity)?.let { activity ->
                    AdFlow.interstitial("global_interstitial").show(
                        activity,
                        object : FullScreenCallback {
                            override fun onAdBlocked(reason: com.adflow.core.BlockReason) = showBlocked(reason.name)
                        },
                    )
                }
            }) { Text("Show Global Interstitial") }

            Button(onClick = {
                (context as? android.app.Activity)?.let { activity ->
                    AdFlow.rewarded("rewarded").show(
                        activity,
                        object : RewardedAdCallback {
                            override fun onUserEarnedReward(reward: RewardItem) { lastReward = reward }
                            override fun onAdBlocked(reason: com.adflow.core.BlockReason) = showBlocked(reason.name)
                        },
                    )
                }
            }) { Text("Show Rewarded Ad") }
            Text("Last reward: ${lastReward?.let { "${it.amount} ${it.type}" } ?: "none yet"}")

            Button(onClick = {
                (context as? android.app.Activity)?.let { activity ->
                    AdFlow.appOpen("app_open").show(activity, FullScreenCallback.EMPTY)
                }
            }) { Text("Show App Open Ad") }

            if (privacyOptionsRequired) {
                Button(onClick = {
                    (context as? android.app.Activity)?.let { AdFlow.consent.showPrivacyOptionsForm(it) {} }
                }) { Text("Privacy options") }
            }

            Button(onClick = { AdFlow.native("home_native").reload() }) {
                Text("Reload Native Ad")
            }

            // AdFlowBanner/AdFlowNative tự quan sát state và tự thay nội dung, kể cả sau reload().
            AdFlowNative(
                placementId = "home_native",
                loading = { ShimmerBox(height = 120.dp) },
                failed = { Text("Native ad unavailable") },
            )
            AdFlowBanner(
                placementId = "home_banner",
                loading = { ShimmerBox(height = 60.dp) },
            )

            val bannerState by AdFlow.banner("home_banner").state.collectAsStateWithLifecycle()
            Text("Banner state: $bannerState")
        }
    }
}

File: flutter/adflow_flutter/example/lib/home_screen.dart
import 'package:adflow_flutter/adflow_flutter.dart';
import 'package:flutter/material.dart';

class HomeScreen extends StatefulWidget {
  const HomeScreen({super.key});
  @override
  State<HomeScreen> createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  bool _premium = false;
  RewardItem? _lastReward;
  bool _privacyOptionsRequired = false;

  @override
  void initState() {
    super.initState();
    _loadPrivacyRequirement();
  }

  Future<void> _loadPrivacyRequirement() async {
    final requirement = await AdFlow.getPrivacyOptionsRequirement();
    if (mounted) {
      setState(
        () => _privacyOptionsRequired =
            requirement == PrivacyOptionsRequirement.required,
      );
    }
  }

  Widget _loading(BuildContext context, double height) => SizedBox(
    height: height,
    child: const Center(child: CircularProgressIndicator()),
  );

  @override
  Widget build(BuildContext context) => Scaffold(
    body: SafeArea(
      child: SingleChildScrollView(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            SwitchListTile(
              contentPadding: EdgeInsets.zero,
              title: const Text('Premium'),
              value: _premium,
              onChanged: (value) {
                setState(() => _premium = value);
                final enabled = !value;
                AdFlow.interstitial('splash_interstitial').setEnabled(enabled);
                AdFlow.interstitial('global_interstitial').setEnabled(enabled);
                AdFlow.appOpen('app_open').setEnabled(enabled);
                AdFlow.banner('home_banner').setEnabled(enabled);
                AdFlow.native('home_native').setEnabled(enabled);
                AdFlow.native('feed_native').setEnabled(enabled);
                AdFlow.native('small_native').setEnabled(enabled);
                // rewarded cố tình không tắt - user chủ động xem để nhận thưởng.
              },
            ),
            FilledButton(
              onPressed: () =>
                  AdFlow.interstitial('global_interstitial').show(),
              child: const Text('Show interstitial'),
            ),
            FilledButton(
              onPressed: () => AdFlow.rewarded('rewarded').show(
                onUserEarnedReward: (reward) =>
                    setState(() => _lastReward = reward),
              ),
              child: const Text('Show rewarded'),
            ),
            Text(
              _lastReward == null
                  ? 'Last reward: none'
                  : 'Last reward: ${_lastReward!.amount} ${_lastReward!.type}',
            ),
            FilledButton(
              onPressed: () => AdFlow.appOpen('app_open').show(),
              child: const Text('Show app open'),
            ),
            if (_privacyOptionsRequired)
              TextButton(
                onPressed: AdFlow.showPrivacyOptionsForm,
                child: const Text('Privacy options'),
              ),
            TextButton.icon(
              onPressed: () => AdFlow.native('home_native').reload(),
              icon: const Icon(Icons.refresh),
              label: const Text('Reload native'),
            ),
            AdFlowNative(
              'home_native',
              loading: (context) => _loading(context, 250),
            ),
            AdFlowNative(
              'feed_native',
              height: 100,
              rendererId: 'compactCard',
              loading: (context) => _loading(context, 100),
            ),
            AdFlowNative(
              'small_native',
              height: 120,
              rendererId: 'small',
              loading: (context) => _loading(context, 120),
            ),
          ],
        ),
      ),
    ),
    bottomNavigationBar: SafeArea(
      child: AdFlowBanner(
        'home_banner',
        loading: (context) => _loading(context, 50),
      ),
    ),
  );
}
